weak_supervision/state_machines/transition_functions/basic_transition_function.py

```diff
@@ -0,0 +1,405 @@
+                                                    new_score,
+                                                    new_rnn_state,
+                                                    considered_actions,
+                                                    probabilities,
+                                                    updated_rnn_state['attention_weights'])
+
+        new_states = []
+        for _, results in batch_action_probs.items():
+            if allowed_actions and not max_actions:
+                # If we're given a set of allowed actions, and we're not just keeping the top k of
+                # them, we don't need to do any sorting, so we can speed things up quite a bit.
+                for group_index, log_probs, _, action_embeddings, actions in results:
+                    for log_prob, action_embedding, action in zip(log_probs, action_embeddings, actions):
+                        if action in allowed_actions[group_index]:
+                            new_states.append(make_state(group_index, action, log_prob, action_embedding))
+            else:
+                # In this case, we need to sort the actions.  We'll do that on CPU, as it's easier,
+                # and our action list is on the CPU, anyway.
+                group_indices = []
+                group_log_probs: List[torch.Tensor] = []
+                group_action_embeddings = []
+                group_actions = []
+                group_curr_log_probs = []
+                for group_index, log_probs, curr_log_probs, action_embeddings, actions in results:
+                    group_indices.extend([group_index] * len(actions))
+                    group_log_probs.append(log_probs)
+                    group_action_embeddings.append(action_embeddings)
+                    group_actions.extend(actions)
+                    group_curr_log_probs.append(curr_log_probs)
+
+
+                log_probs = torch.cat(group_log_probs, dim=0)
+                curr_log_probs = torch.cat(group_curr_log_probs, dim=0)
+                action_embeddings = torch.cat(group_action_embeddings, dim=0)
+                log_probs_cpu = log_probs.data.cpu().numpy().tolist()
+                batch_states = [(log_probs_cpu[i],
+                                 group_indices[i],
+                                 log_probs[i],
+                                 action_embeddings[i],
+                                 group_actions[i])
+                                for i in range(len(group_actions))
+                                if (not allowed_actions or
+                                    group_actions[i] in allowed_actions[group_indices[i]])]
+                # this is the only change required to sample from current log probabilities instead of beam search
+                if sample:
+                    categorical_dist = Categorical(logits=curr_log_probs)
+                    _, group_index, log_prob, action_embedding, action = batch_states[categorical_dist.sample()]
```

Comment:
Looks like you're sampling only one action here. Was that intentional?

Reply:
I guess it was. So the idea is that you sample one action at each time step instead of keeping a sorted beam, correct?

Reply:
Yes, so I want a sampler function that can sample a whole decoded sequence, and then call THAT multiple times. So at the end i'll have max_num_decoded_sequences number of sequences for each batch instance